File: 2001-3000/2875-minimum-size-subarray-in-infinite-array/2875-minimum-size-subarray-in-infinite-array.java
class Solution {
    public int minSizeSubarray(int[] n, int t) {
        int b=0;
        
        long x = 0;
        for(int i=0; i<n.length; i++)x+=n[i];

        if(x<=t){
            b+=n.length*(t/x);
            t%=(int)x;
        }

        if(t==0)return b;

        int[] a1=get(n, t);
        int[] a2=get(n, x-t);

        if(a1[0]>n.length&&a2[1]<0)return -1;
        else if(a1[0]>n.length)return b+n.length-a2[1];
        else if(a2[1]<0)return b+a1[0];
        else return b+Math.min(a1[0], n.length-a2[1]);
    }
    private int[] get(int[] n, long v){
        int[] a = new int[]{2_000_000_000, -2_000_000_000};
        long m=0;
        int j=0;
        for(int i=0; i<n.length; i++){
            for(; j<n.length&&m<v; j++)m+=n[j];
            if(m<v)return a;
            if(m==v){
                a[0]=Math.min(a[0], j-i);
                a[1]=Math.max(a[1], j-i);
            }
            m-=n[i];
        }
        return a;
    }
}

File: 2001-3000/2875-minimum-size-subarray-in-infinite-array/2875-minimum-size-subarray-in-infinite-array.cpp
class Solution {
public:
    int minSizeSubarray(vector<int>& nums, int target) {

        long long total = 0;
        int n = nums.size();

        for (int x : nums)
            total += x;

        long long full = target / total;
        long long rem = target % total;

        if (rem == 0)
            return full * n;

        long long sum = 0;
        int i = 0;
        int ans = INT_MAX;

        for (int j = 0; j < 2 * n; j++) {

            sum += nums[j % n];

            while (sum > rem) {
                sum -= nums[i % n];
                i++;
            }

            if (sum == rem) {
                ans = min(ans, j - i + 1);
            }
        }

        if (ans == INT_MAX)
            return -1;

        return full * n + ans;
    }
};